src/ast.rs

```diff
@@ -1436,8 +1447,8 @@ impl Parse for Encoding {
     ) -> Result<(Encoding, IndexStr<'b>)> {
         try_begin_parse!("Encoding", ctx, input);
 
-        if let Ok((name, tail)) = Name::parse(ctx, subs, input) {
-            if let Ok((ty, tail)) = BareFunctionType::parse(ctx, subs, tail) {
+        if let Ok((name, tail)) = try_recurse!(Name::parse(ctx, subs, input)) {
+            if let Ok((ty, tail)) = try_recurse!(BareFunctionType::parse(ctx, subs, tail)) {
```

Comment:
I guess another option would be some sort of helper like

```rust
fn parse<'a, 'b, P: Parse>(ctx: &'a ParseContext, subs: &'a mut SubstitutionTable, input: IndexStr<'b>) -> Result<Result<(P, IndexStr<'b>)>> {
    match P::parse(ctx, subs, input) {
        Err(Error::TooMuchRecursion) => Err(Error::TooMuchRecursion),
        x => Ok(x),
    }
}
```

and then we could use it like

```rust
if let Ok((name, tail)) = parse::<Name>(ctx, subs, input)? {
    // ...
}
```

This is a little less macro-y and lets us use `?`.

What do you think of that?

Reply:
I'm not really a fan of this because there are many sites where all Foo::parse errors are propagated immediately, and then it's not clear where you need to use this helper and where you don't.

I think what we really want (and what doesn't exist) is a tri-state Result with Ok/Recoverable/Irrecoverable errors and two ? operators one that propagates Irrecoverable only and one that propagates both.

Reply:
> I think what we really want (and what doesn't exist) is a tri-state Result with Ok/Recoverable/Irrecoverable errors and two ? operators one that propagates Irrecoverable only and one that propagates both.

I guess I was imagining that `Result<Result<T>>` would be that tri-state where

* `Ok(Ok(_))` was success
* `Ok(Err(_))` was recoverable error
* and `Err(_)` was irrecoverable error

But again, no strong opinions!